contracts/utils/structs/CircularBuffer.sol

```diff
@@ -0,0 +1,114 @@
+     * where the next value should be stored.
+     *
+     * Struct members have an underscore prefix indicating that they are "private" and should not be read or written to
+     * directly. Use the functions provided below instead. Modifying the struct manually may violate assumptions and
+     * lead to unexpected behavior.
+     *
+     * The last item is at data[(index - 1) % data.length] and the last item is at data[index % data.length]. This
+     * range can wrap around.
+     */
+    struct Bytes32CircularBuffer {
+        uint256 _count;
+        bytes32[] _data;
+    }
+
+    /**
+     * @dev Initialize a new CircularBuffer of given length.
+     *
+     * If the CircularBuffer was already setup and used, calling that function again will reset it to a blank state.
+     */
+    function setup(Bytes32CircularBuffer storage self, uint256 length) internal {
+        clear(self);
+        Arrays.unsafeSetLength(self._data, length);
+    }
+
+    /**
+     * @dev Clear all data in the buffer, keeping the existing length.
+     */
+    function clear(Bytes32CircularBuffer storage self) internal {
+        self._count = 0;
+    }
+
+    /**
+     * @dev Push a new value to the buffer. If the buffer is already full, the new value replaces the oldest value in
+     * the buffer.
+     */
+    function push(Bytes32CircularBuffer storage self, bytes32 value) internal {
+        uint256 index = self._count++;
+        uint256 length = self._data.length;
+        Arrays.unsafeAccess(self._data, index % length).value = value;
+    }
+
+    /**
+     * @dev Number of values currently in the buffer. This values is 0 for empty buffer, and cannot exceed the size of
+     * the buffer.
+     */
+    function count(Bytes32CircularBuffer storage self) internal view returns (uint256) {
+        return Math.min(self._count, self._data.length);
+    }
+
+    /**
+     * @dev Length of the buffer. This is the maximum number of elements kepts in the buffer.
+     */
+    function size(Bytes32CircularBuffer storage self) internal view returns (uint256) {
+        return self._data.length;
+    }
+
+    /**
+     * @dev Getter for the i-th value in the buffer, from the end.
+     *
+     * Reverts with {Panic-ARRAY_OUT_OF_BOUNDS} if trying to access an element that was not pushed, or that was
+     * dropped to make room for newer elements.
+     */
+    function last(Bytes32CircularBuffer storage self, uint256 i) internal view returns (bytes32) {
+        uint256 index = self._count;
+        uint256 length = self._data.length;
+        uint256 total = Math.min(index, length); // count(self)
+        if (i >= total) {
+            Panic.panic(Panic.ARRAY_OUT_OF_BOUNDS);
+        }
+        return Arrays.unsafeAccess(self._data, (index - i - 1) % length).value;
+    }
+
+    /**
+     * @dev Check if a given value is in the buffer.
+     */
+    function includes(Bytes32CircularBuffer storage self, bytes32 value) internal view returns (bool) {
```

Comment:
This function is an unbounded loop. Shall we restrict the maximum length of the buffer?
The alternative may be just a warning

Reply:
The loop is from i = 0 to total, with total being `Math.min(index, length)` ... so it is bounded by the length of the buffer already

Reply:
Yes, but the buffer can be huge. This wasn't a problem in Merkle Tree originally because we restricted the maximum length to 255. Perhaps we should do the same here.

Reply:
You asked "Shall we restrict the maximum length of the buffer?". I' saying it already is.

Now I understand you are concerned this might be too large. Maybe. That is already documented at the top of the contract when we say that the complexity is ` * - inclusion ({includes}): O(N) (worst case)`

Reply:
Fair, I'm just worried it's easy to make a mistake when choosing the size of the buffer even if the NatSpec mentions its complexity. 

Also, it's fair that CircularBuffer users might not need the `includes` function, otherwise, I may recommend using an EnumerableSet instead.

I'll clarify at the `setup` function that a selecting a `size` may render the `includes` function unusable